docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,268 @@
+or generate them itself. In the first case, this is a burden on the cluster
+admin and a significant logistical barrier to secure Kubernetes rollouts. In
+the second, the kubelet must self-sign its certificate and forfeits many of the
+advantages of a PKI system. Instead, we propose that the kubelet generate a
+private key and a CSR for submission to a cluster-level certificate signing
+process.
+
+## Preliminaries
+
+We assume the existence of a functioning control plane. The
+apiserver should be configured for TLS initially or possess the ability to
+generate valid TLS credentials for itself. If secret information is passed in
+the request (e.g. auth tokens supplied with the request or included in
+ExtraInfo) then all communications from the node to the apiserver must take
+place over a verified TLS connection.
+
+Each node is additionally provisioned with the following information:
+
+1. Location of the apiserver
+2. Any CA certificates necessary to trust the apiserver's TLS certificate
+3. Access tokens (if needed) to communicate with the CSR endpoint
+
+These should not change often and are thus simple to include in a static
+provisioning script.
+
+## API Changes
+
+### CertificateSigningRequest Object
+
+We introduce a new API object to represent PKCS#10 certificate signing
+requests. It will be accessible under:
+
+`/apis/certificates/v1beta1/signingrequests/mycsr`
+
+It will have the following structure:
+
+```go
+// Describes a certificate signing request
+type CertificateSigningRequest struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// The certificate request itself and any additonal information.
+	Spec CertificateSigningRequestSpec `json:"spec,omitempty"`
+
+	// Derived information about the request.
+	Status CertificateSigningRequestStatus `json:"status,omitempty"`
+
+	// The current approval state of the request.
+	Approve CertificateSigningRequestApproval `json:"approve,omitempty"`
+}
+
+// This information is immutable after the request is created.
+type CertificateSigningRequestSpec struct {
+	// base64-encoded PKCS#10 CSR data
+	CertificateRequest string `json:"request"`
+
+	// Any extra information the node wishes to send with the request.
+	ExtraInfo []string `json:"extra,omitempty"`
+}
+
+// This information is derived from the request by Kubernetes and cannot be
+// modified by users. All information is optional since it might not be
+// available in the underlying request. This is intented to aid approval
+// decisions.
+type CertificateSigningRequestStatus struct {
+	// Information about the requesting user (if relevant)
+	// See user.Info interface for details
+	Username string   `json:"username,omitempty"`
+	UID      string   `json:"uid,omitempty"`
+	Groups   []string `json:"groups,omitempty"`
+
+	// Fingerprint of the public key in request
+	Fingerprint string `json:"fingerprint,omitempty"`
+
+	// Subject fields from the request
+	Subject pkix.Name `json:"subject,omitempty"`
```

Comment:
what is a pkix.Name? we should import a representation of it into the api so we don't depend on non api packages.